src/java/org/apache/cassandra/db/tries/SlicedTrie.java

```diff
@@ -61,151 +61,279 @@ public SlicedTrie(Trie<T> source, ByteComparable left, boolean includeLeft, Byte
         this.includeRight = includeRight;
     }
 
+    static ByteSource add0(ByteSource src)
+    {
+        return new ByteSource()
+        {
+            boolean done = false;
+
+            @Override
+            public int next()
+            {
+                if (done)
+                    return END_OF_STREAM;
+                int next = src.next();
+                if (next != END_OF_STREAM)
+                    return next;
+                done = true;
+                return 0;
+            }
+        };
+    }
```

Comment:
I've noticed this pattern elsewhere. Here is one such place:

https://github.com/datastax/cassandra/blob/f679bce038fabc2a0b69c74441a4b01121bd95b0/src/java/org/apache/cassandra/index/sai/memory/TrieMemoryIndex.java#L271-L290

It's not a huge rush, but it seems like a generic class that could be added as a peer to `ByteSource.withTerminator`. That method uses `Multi`, which doesn't give us the same semantics as seen in this method and the one I reference above, so we would need to make it clear that we're simply adding a terminator.